runtime/lua/vim/lsp/document_color.lua

```diff
@@ -0,0 +1,296 @@
+    or not bufstate.enabled
+  then
+    return
+  end
+
+  if not bufstate.hl_infos or not bufstate.buf_version then
+    bufstate.hl_infos = {}
+    bufstate.buf_version = ctx.version
+  end
+
+  local color_ranges = vim
+    .iter(result)
+    --- @param res lsp.ColorInformation
+    :map(function(res)
+      local range = {
+        res.range.start.line,
+        res.range.start.character,
+        res.range['end'].line,
+        res.range['end'].character,
+      }
+      return { hl_group = get_hl_group(res.color), range = range }
+    end)
+    :totable()
+  bufstate.hl_infos[client_id] = color_ranges
+
+  bufstate.buf_version = ctx.version
+  api.nvim__redraw({ buf = bufnr, valid = true, flush = false })
+end
+
+--- Query whether document colors are enabled in the filtered scope.
+---
+--- @param filter? lsp.document_color.enable.Filter
+--- @return boolean
+function M.is_enabled(filter)
+  vim.validate('filter', filter, 'table', true)
+
+  filter = filter or {}
+  local bufnr = filter.bufnr
+
+  if bufnr == nil then
+    return global_state.enabled
+  else
+    return bufstates[vim._resolve_bufnr(bufnr)].enabled
+  end
+end
+
+--- Optional filters |kwargs|, or `nil` for all.
+--- @class lsp.document_color.enable.Filter
+--- @inlinedoc
+--- Buffer number, or 0 for current buffer, or nil for all.
+--- @field bufnr? integer
+
+--- Enables document highlighting from the given language client in the given buffer.
+---
+--- To "toggle", pass the inverse of `is_enabled()`:
+---
+--- ```lua
+--- lsp.document_color.enable(not vim.lsp.document_color.is_enabled())
+--- ```
+---
+--- @param enable? boolean True to enable, false to disable
+--- @param filter? lsp.document_color.enable.Filter
+function M.enable(enable, filter)
+  vim.validate('enable', enable, 'boolean', true)
+  vim.validate('filter', filter, 'table', true)
+
+  enable = enable == nil or enable
+  filter = filter or {}
+
+  if filter.bufnr == nil then
+    global_state.enabled = enable
+    for _, bufnr in ipairs(api.nvim_list_bufs()) do
+      if api.nvim_buf_is_loaded(bufnr) then
+        if enable then
+          buf_enable(bufnr)
+        else
+          buf_disable(bufnr)
+        end
+      else
+        bufstates[bufnr] = nil
+      end
+    end
+  else
+    local bufnr = vim._resolve_bufnr(filter.bufnr)
+    if enable then
+      buf_enable(bufnr)
+    else
+      buf_disable(bufnr)
+    end
+  end
+end
+
+api.nvim_create_autocmd('LspNotify', {
+  group = document_color_augroup,
+  callback = function(args)
+    local bufnr = args.buf
+    local method = args.data.method --- @type string
+
+    if
+      (method == ms.textDocument_didChange or method == ms.textDocument_didOpen)
+      and bufstates[bufnr].enabled
+    then
+      buf_refresh(bufnr, { client_id = args.data.client_id })
+    end
+  end,
+})
+
+api.nvim_create_autocmd('LspAttach', {
+  group = document_color_augroup,
+  callback = function(args)
+    api.nvim_buf_attach(args.buf, false, {
```

Comment:
Is it possible to not unconditionally attach to every LSP enabled buffer? Shouldn't this only be created when `enable` is called?

Reply:
Hmm I borrowed the logic from [`inlay_hint`](https://github.com/neovim/neovim/blob/1de276bbcd8f1b3c5ef1d0bd6d4e112cae119337/runtime/lua/vim/lsp/inlay_hint.lua#L267) but yes you're right we can probably improve this.